man/orm.1
.Dd February 19, 2025
.Dt ORM 1
.Os
.Sh NAME
.Nm orm
.Nd jormungandr iterative build sandbox
.Sh SYNOPSIS
.Nm orm
.Op Fl PSUir
.Op Fl t Ar toolchain
.Op Fl b Ar bsys
.Op Fl w Ar workspace
.Op Fl u Ar sysroot
.Op Fl d Ar destdir
.Op Fl o Ar objdir
.Op Fl s Ar srcdir
.Op Ar arguments ...
.Nm orm
.Op Fl P
.Op Fl w Ar workspace
.Op Fl s Ar srcdir
.Fl p Ar workdir
.Nm orm
.Fl h
.Sh DESCRIPTION
Execute a build system driver script
.Pq Ar bsys
inside a sandbox dedicated for source code builds.
.Bl -tag
.It Fl P
Use persistent build directories. By default,
.Nm
workdirs are located in the user's session runtime directory,
.Po based on
.Ev XDG_RUNTIME_DIR Pc
When this option is set,
.Nm
uses the user's cache directory
.Po based on
.Ev XDG_CACHE_HOME Pc
to create its default
.Ar destdir
and
.Ar objdir .
.It Fl S
Mount
.Ar srcdir
with write-access. By default,
.Ar srcdir
is mounted read-only in the sandbox. If the build system
doesn't support out-of-tree builds, this option allows
write-access. Note that most modern build systems do support
out-of-tree builds, and you should create your driver to
support this. This option is only available as a last resort.
.It Fl U
Mount
.Ar sysroot
with write-access. By default,
.Ar sysroot
is mounted read-only in the sandbox. If you are creating
a system image on-the-fly, you might want to directly export
some libraries and headers.
.It Fl i
Start an interactive shell in the sandbox, for manual experimentation, debug, etc...
.It Fl r
Usurpate 0:0
.Pq root's
user and group id instead of the default 1000:1000 credentials in the sandbox.
.It Fl t Ar toolchain
Specify the
.Ar toolchain
mounted read-only as the sandbox root directory.
.It Fl b Ar bsys
Specify the
.Ar bsys
made available read-only in
.Pa /var/bsys
and executed if not interactive
.Pq no Fl i .
.It Fl w Ar workspace
Specify the
.Ar workspace
in which
.Nm
workdirs are created, if none specified, it is inferred from
.Ar srcdir .
.It Fl u Ar sysroot
Specify the
.Ar sysroot
directory. Read-only by default.
.It Fl d Ar destdir
Specify the
.Ar destdir
where files should be staged for packaging. Never read-only.
.It Fl o Ar objdir
Specify the
.Ar objdir
where object files and other build artifacts are created.
.It Fl s Ar srcdir
Specify the
.Ar srcdir
source code directory. Read-only by default.
.It Ar arguments ...
Arguments forwarded to the
.Ar bsys
or the shell if interactive
.Pq Fl i .
.It Fl p Ar workdir
In this scenario, prints the absolute path of the given
.Ar workdir
associated with the current
.Ar workspace
or
.Ar srcdir
.Pq and persistence if Fl P .
.It Fl h
Print usage and exit.
.Sh ENVIRONMENT
.Bl -tag
.It Ev ORM_SRCDIR_COMMAND
Set the
.Ar srcdir
resolution command.
.It Ev ORM_DEFAULT_TOOLCHAIN
Set the default
.Ar toolchain
to use.
.It Ev ORM_DEFAULT_BSYS
Set the default
.Ar bsys
to use.
.It Ev ORM_WORKSPACE
Set the
.Ar workspace
name to use if none specified.
.It Ev ORM_SYSROOT
Set the
.Ar sysroot
to mount if none specified.
.Sh EXIT STATUS
The
.Nm
utility exits with the
.Ar bsys
return value on success setting up and entering
the sandbox, and >0 if an error occurs before that.
.Sh SEE ALSO
.Xr lndworm 1 , Xr gitworm 1 .
.Sh AUTHORS
Written by
.An Valentin Debon Aq Mt valentin.debon@heylelos.org .


src/orm.c
/* SPDX-License-Identifier: AGPL-3.0-or-later */
#include <stdio.h> /* puts, fprintf */
#include <stdlib.h> /* realpath, ... */
#include <stdnoreturn.h> /* noreturn */
#include <string.h> /* strdup, memcpy, ... */
#include <unistd.h> /* getopt, ... */
#include <libgen.h> /* dirname, basename */
#include <alloca.h> /* alloca */
#include <err.h> /* err, errx, warnx */

#include <orm.h>

#include "common/cmdpath.h"

struct orm_args {
	const char *toolchain, *bsys;
	const char *workspace, *sysroot;
	const char *destdir, *objdir, *srcdir;
	const char *workdir;
	unsigned int asroot : 1, rwsysroot : 1, rwsrcdir : 1;
	unsigned int persistent : 1, interactive : 1;
};

/**
 * Finds out and prints the absolute path of the given
 * workdir in the resolved (or given) workspace.
 * @param args Command line options.
 * @return Never
 */
noreturn static void
orm_print_workdir(const struct orm_args *args) {
	char *workdir;
	int flags = 0;

	if (args->persistent) {
		flags |= ORM_WORKDIR_PERSISTENT;
	}

	if (orm_workdir(args->workspace, args->workdir, flags, &workdir) != 0) {
		err(EXIT_FAILURE, "Unable to lookup workdir '%s'", args->workdir);
	}

	puts(workdir);

	exit(EXIT_SUCCESS);
}

/**
 * Execute the given bsys, or the user's shell if available.
 * @param bsysname The base name of the bsys, or NULL if interactive.
 * @param args Arguments to forward to the command.
 * @param count Number of arguments in args.
 * @return Never
 */
noreturn static void
orm_exec(const char *bsysname, char **args, int count) {
	char *argv[3 + count];
	int argc = 0;

	if (bsysname != NULL) {
		static const char bsysdir[] = "/var/bsys/";
		const size_t bsysnamelen = strlen(bsysname);
		char * const bsys = alloca(sizeof (bsysdir) + bsysnamelen);

		memcpy(bsys, bsysdir, sizeof (bsysdir) - 1);
		memcpy(bsys + sizeof (bsysdir) - 1, bsysname, bsysnamelen + 1);

		argv[argc++] = bsys;
	} else {
		char *shell = getenv("SHELL");

		if (shell == NULL || access(shell, X_OK) != 0) {
			shell = "/bin/sh";
		}
		argv[argc++] = shell;
		argv[argc++] = "-i";
	}

	while (count != 0) {
		argv[argc++] = *args++;
		count--;
	}
	argv[argc] = NULL;

	execv(*argv, argv);

	err(EXIT_FAILURE, "execv '%s'", *argv);
}

/**
 * Setup the sandbox and execute the bsys, or go interactive.
 * @param args Command line options.
 * @param argc Forwarded from main().
 * @param argv Forwarded from main().
 * @return Never
 */
noreturn static void
orm_run(const struct orm_args *args, int argc, char **argv) {
	struct orm_sandbox_description description = {
		.sysroot = args->sysroot, .destdir = args->destdir,
		.objdir = args->objdir, .srcdir = args->srcdir,
		.asroot = args->asroot, .rosysroot = !args->rwsysroot,
		.rosrcdir = !args->rwsrcdir,
	};

	/* Use persistent-cache if requested. */
	int flags = 0;
	if (args->persistent) {
		flags |= ORM_WORKDIR_PERSISTENT;
	}

	if (description.objdir == NULL) {
		char *objdir;
		if (orm_workdir(args->workspace, "obj", flags, &objdir) != 0) {
			err(EXIT_FAILURE, "Unable to lookup objdir");
		}
		description.objdir = objdir;
	}

	if (description.destdir == NULL) {
		char *destdir;
		if (orm_workdir(args->workspace, "dest", flags, &destdir) != 0) {
			err(EXIT_FAILURE, "Unable to lookup destdir");
		}
		description.destdir = destdir;
	}

	/* Resolve the toolchain's path. */
	char *root;
	if (orm_toolchain_path(args->toolchain, &root) != 0) {
		err(EXIT_FAILURE, "Unable to find toolchain '%s'", args->toolchain);
	}
	description.root = root;

	/* If bsys is a path (relative or not), use it directly,
	 * else, search the user's configurations for our bsys. */
	char *bsyspath;
	if (strchr(args->bsys, '/') != NULL) {
		bsyspath = strdup(args->bsys);
	} else if (orm_bsys_path(args->bsys, &bsyspath) != 0) {
		err(EXIT_FAILURE, "Unable to find bsys '%s'", args->bsys);
	}
	description.bsysdir = dirname(strdupa(bsyspath));

	char *bsysname;
	if (!args->interactive) {
		bsysname = basename(strdupa(bsyspath));
	} else {
		bsysname = NULL;
	}

	/* Enter the sandbox, as we don't need anything from the system now. */
	if (orm_sandbox(&description, getuid(), getgid()) != 0) {
		err(EXIT_FAILURE, "Unable to enter toolbox");
	}

	/* Execute either the bsys or the shell. */
	orm_exec(bsysname, argv + optind, argc - optind);
}

noreturn static void
orm_usage(const char *progname, int status) {

	fprintf(stderr,
		"usage: %1$s [-PSUir] [-t <toolchain>] [-b <bsys>] [-w <workspace>] [-u <sysroot>] [-d <destdir>] [-o <objdir>] [-s <srcdir>] [<arguments>...]\n"
		"       %1$s [-P] [-w <workspace>] [-s <srcdir>] -p <workdir>\n"
		"       %1$s -h\n",
		progname);

	exit(status);
}

static struct orm_args
orm_parse_args(int argc, char **argv) {
	struct orm_args args = {
		.toolchain = getenv("ORM_DEFAULT_TOOLCHAIN"),
		.bsys = getenv("ORM_DEFAULT_BSYS"),
		.workspace = getenv("ORM_WORKSPACE"),
		.sysroot = getenv("ORM_SYSROOT"),
	};
	int c;

	while ((c = getopt(argc, argv, ":hPSUirt:b:w:u:d:o:s:p:")) >= 0) {
		switch (c) {
		case 'h': orm_usage(*argv, EXIT_SUCCESS);
		case 'P': args.persistent = 1; break;
		case 'S': args.rwsrcdir = 1; break;
		case 'U': args.rwsysroot = 1; break;
		case 'i': args.interactive = 1; break;
		case 'r': args.asroot = 1; break;
		case 't': args.toolchain = optarg; break;
		case 'b': args.bsys = optarg; break;
		case 'w': args.workspace = optarg; break;
		case 'u': args.sysroot = optarg; break;
		case 'd': args.destdir = optarg; break;
		case 'o': args.objdir = optarg; break;
		case 's': args.srcdir = optarg; break;
		case 'p': args.workdir = optarg; break;
		case ':':
			warnx("Option -%c requires an operand", optopt);
			orm_usage(*argv, EXIT_FAILURE);
		case '?':
			warnx("Unrecognized option -%c", optopt);
			orm_usage(*argv, EXIT_FAILURE);
		}
	}

	/* No need to resolve a srcdir if we are just
	 * here for a workdir and workspace is already specified.
	 * This allows this synopsis to run without executing srccmd. */
	if (args.workdir == NULL || args.workspace == NULL) {
		/* Resolve the given (or not) source directory into
		 * an absolute path, either with cmdpath() or realpath(3). */
		if (args.srcdir == NULL) {
			const char *srccmd = getenv("ORM_SRCDIR_COMMAND");

			if (srccmd == NULL) {
				srccmd = CONFIG_DEFAULT_SRCDIR_COMMAND;
			}

			args.srcdir = cmdpath(srccmd);
		} else {
			args.srcdir = realpath(args.srcdir, NULL);
		}

		if (args.srcdir == NULL) {
			warn("Unable to lookup srcdir");
			orm_usage(*argv, EXIT_FAILURE);
		}
	}

	/* Resolve the workspace, as a convenience,
	 * if none is specified, the srcdir's basename is used. */
	if (args.workspace == NULL) {
		args.workspace = strdup(basename(strdupa(args.srcdir)));
	}

	/* Thus we need to check the workspace is a valid candidate... */
	if (args.workspace == NULL || *args.workspace == '\0'
		|| *args.workspace == '.' || strcmp(args.workspace, "/") == 0) {
		warnx("Unable to find out workspace name");
		orm_usage(*argv, EXIT_FAILURE);
	}

	if (args.toolchain == NULL) {
		args.toolchain = "default";
	}

	if (args.bsys == NULL) {
		args.bsys = "default";
	}

	if (args.sysroot == NULL) {
		args.sysroot = "/";
	}

	return args;
}

int
main(int argc, char *argv[]) {
	const struct orm_args args = orm_parse_args(argc, argv);

	if (args.workdir != NULL) {
		orm_print_workdir(&args);
	}

	orm_run(&args, argc, argv);
}
